### PR TITLE
[Glimmer2] hasBlock support

### PR DESCRIPTION
**Pull request by Joelkang**
Added tests for the variety of ways `has-block` can be used in glimmer

### CODE REVIEW

File: packages/ember-glimmer/lib/ember-template-compiler/plugins/transform-has-block-syntax.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  @method transform
+  @param {AST} ast The AST to be transformed.
+*/
+TransformHasBlockSyntax.prototype.transform = function TransformHasBlockSyntax_transform(ast) {
+  let { traverse, builders: b } = this.syntax;
+
+  traverse(ast, {
+    PathExpression(node) {
+      if (node.original === OLD_HAS_BLOCK) {
+        return b.sexpr(b.path(NEW_HAS_BLOCK));
```
Comment (mmun) [inline packages/ember-glimmer/lib/ember-template-compiler/plugins/transform-has-block-syntax.js]: We should pass through the `loc` of the PathExpression into the SubExpression. Even though it is kind of hacky, it still represents the source code that ultimately generated the subexpression. Same applies below in the SubExpression visitor.

**Comment by mmun** [comment]
@Joelkang Excellent!

File: packages/ember-glimmer/lib/ember-template-compiler/plugins/transform-has-block-syntax.js (hunk 1 above)
Comment (Joelkang) [inline packages/ember-glimmer/lib/ember-template-compiler/plugins/transform-has-block-syntax.js]: Looks like buildSexpr and buildPath [don't actually support](https://github.com/tildeio/glimmer/blob/master/packages/glimmer-syntax/lib/builders.ts#L93) a `loc` parameter. I could add it to glimmer and then update this PR?

**Comment by Joelkang** [comment]
Much thanks @mmun, replied to the comment; also let me know if hasBlockParams should go into a separate PR.

Incidentally, I'm not entirely sure what `{{hasBlockParams "inverse"}}` actually does (but I added tests for it and it transforms correctly)

File: packages/ember-glimmer/lib/ember-template-compiler/plugins/transform-has-block-syntax.js (hunk 1 above)
Comment (mmun) [inline packages/ember-glimmer/lib/ember-template-compiler/plugins/transform-has-block-syntax.js]: Ah! Let's merge this as is and fix that separately then.

**Comment by mmun** [comment]
@Joelkang `{{hasBlockParams "inverse"}}` is always false because it is currently not syntactically possible to use block params in an inverse block. Keeping the test around seems fine though.

**Comment by Joelkang** [comment]
Fascinating, looks like htmlbars will return `1` or `0` when `{{hasBlockParams}}` is used in an attribute without a named block. `<button name={{hasBlockParams "default"}}` actually returns `<button name="true">` Marking as `@glimmer` for now

**Comment by mmun** [comment]
Interesting. It must be that it returns the `arity` of the block rather than `!!arity`?

**Comment by Joelkang** [comment]
[Looks to be the case](https://github.com/emberjs/ember.js/blob/master/packages/ember-htmlbars/lib/hooks/get-root.js#L13), but I'm not familiar enough with the HTMLBars code to say if we should fix this.

It'll probably be a regression but then again I'm hoping no one is actually relying on the output being 1 or 0 rather than a boolean; if it's being used in purely logical manners, `!!arity` probably won't break anyone. Not sure what the guidance is on something like that though, so I'll defer to you on whether we should fix it or leave it.

**Comment by rwjblue** [comment]
It could be used to do different things based on the number of block params provided. I am not sure that it is terribly important to preserve (AFAIK this was not documented or publicized).

**Comment by mmun** [comment]
:+1: `!!arity`.

**Comment by chadhietala** [comment]
This LGTM.

**Pull request merged**